clippy_lints/src/octal_escapes.rs

```diff
@@ -0,0 +1,131 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use rustc_ast::ast::{Expr, ExprKind};
+use rustc_ast::token::{Lit, LitKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `\0` escapes in string and byte literals that look like octal character
+    /// escapes in C
+    ///
+    /// ### Why is this bad?
+    /// Rust does not support octal notation for character escapes. `\0` is always a
+    /// null byte/character, and any following digits do not form part of the escape
+    /// sequence.
+    ///
+    /// ### Known problems
+    /// The actual meaning can be the intended one. `\x00` can be used in these
+    /// cases to be unambigious.
+    ///
+    /// # Example
+    /// ```rust
+    /// // Bad
+    /// let one = "\033[1m Bold? \033[0m";  // \033 intended as escape
+    /// let two = "\033\0";                 // \033 intended as null-3-3
+    ///
+    /// // Good
+    /// let one = "\x1b[1mWill this be bold?\x1b[0m";
+    /// let two = "\x0033\x00";
+    /// ```
+    #[clippy::version = "1.58.0"]
+    pub OCTAL_ESCAPES,
+    suspicious,
+    "string escape sequences looking like octal characters"
+}
+
+declare_lint_pass!(OctalEscapes => [OCTAL_ESCAPES]);
+
+impl EarlyLintPass for OctalEscapes {
+    fn check_expr(&mut self, cx: &EarlyContext<'tcx>, expr: &Expr) {
+        if in_external_macro(cx.sess, expr.span) {
+            return;
+        }
+
+        if let ExprKind::Lit(lit) = &expr.kind {
+            if matches!(lit.token.kind, LitKind::Str) {
+                check_lit(cx, &lit.token, lit.span, true);
+            } else if matches!(lit.token.kind, LitKind::ByteStr) {
+                check_lit(cx, &lit.token, lit.span, false);
+            }
+        }
+    }
+}
+
+fn check_lit(cx: &EarlyContext<'tcx>, lit: &Lit, span: Span, is_string: bool) {
+    let contents = lit.symbol.as_str();
+    let mut iter = contents.char_indices();
+
+    // go through the string, looking for \0[0-7]
+    while let Some((from, ch)) = iter.next() {
+        if ch == '\\' {
+            if let Some((mut to, '0')) = iter.next() {
+                // collect all further potentially octal digits
+                while let Some((j, '0'..='7')) = iter.next() {
+                    to = j + 1;
+                }
+                // if it's more than just `\0` we have a match
+                if to > from + 2 {
+                    emit(cx, &contents, from, to, span, is_string);
+                    return;
+                }
```

Comment:
I think that this check is a bit more complicated than it has to be, with the `j + 1`. You can drop that and adjust the check accordingly. Maybe like this:
```rs
while let Some((j, '0'..='7')) = iter.next() {
    to = j;
}
// if it's more than just `\0` we have a match
if (to - from) > 1 {
    emit(cx, &contents, from, to, span, is_string);
    return;
}
```

Reply:
Ok, I'll simplify this.